events: ensure the name is valid

aojea opened the pull request:
The event Object is created from the referenced Object name, however, there is no guarantee that the name from the referenced Object will be a valid Event Name.

For those Objects with name not valid for an Event, use the base64 representation of the name without padding.

/kind bug
Fixes #127594

```release-note
bugfix for Events that fail to be created when the referenced object name is not a valid Event name, using an UUID as name instead of the referenced object name and the timestamp suffix.
```

aojea commented:
/assign @dgrisonnet @liggitt

aojea commented:
/sig api-machinery

liggitt commented on staging/src/k8s.io/client-go/tools/events/event_recorder.go:
```diff
@@ -94,9 +96,18 @@ func (recorder *recorderImpl) makeEvent(refRegarding *v1.ObjectReference, refRel
+	// Ref: https://issues.k8s.io/127594
+	name := fmt.Sprintf("%s.%x", refRegarding.Name, t.UnixNano())
+	if len(apimachineryvalidation.NameIsDNSSubdomain(name, false)) > 0 {
+		name = fmt.Sprintf("%s.%x", base64.RawURLEncoding.EncodeToString([]byte(refRegarding.Name)), t.UnixNano())
```
if the involved object name is long, do we need to possibly trim the length of the encoded segment to fit within the max? something like this (completely untested, might be off by one):

```go
suffix := fmt.Sprintf(".%x", t.UnixNano())
prefix := base64.RawURLEncoding.EncodeToString([]byte(refRegarding.Name))
if len(prefix)+len(suffix) > apimachineryvalidation.DNS1123SubdomainMaxLength {
  prefix = prefix[:apimachineryvalidation.DNS1123SubdomainMaxLength-len(suffix)]
}
name = prefix + suffix
```

liggitt replied:
also, I think it's possible for url-base64-encoding to start or end with `-`, or contain _, both of which I think would fail event name validation?

aojea replied:
> also, I think it's possible for url-base64-encoding to start or end with `-`, or contain _, both of which I think would fail event name validation?

you. make me think hard, but I think is not possible if it is derived from an alphanumeric character https://www.ascii-code.com/ based on the encoding https://datatracker.ietf.org/doc/html/rfc4648#section-5 , but is simpler to implement a trim those chars than trying to resolve that puzzle 😅

aojea replied:
see latest commit to not use base64 https://github.com/kubernetes/kubernetes/pull/129790#discussion_r1931002836

aojea commented:
/retest

failure is https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129790/pull-kubernetes-e2e-gce/1882832312173858816

liggitt commented on staging/src/k8s.io/client-go/tools/events/event_recorder.go:
```diff
@@ -94,9 +98,23 @@ func (recorder *recorderImpl) makeEvent(refRegarding *v1.ObjectReference, refRel
+		if len(prefix)+len(suffix) > utilvalidation.DNS1123SubdomainMaxLength {
+			prefix = prefix[:utilvalidation.DNS1123SubdomainMaxLength-len(suffix)]
+		}
+		name = strings.Trim(prefix+suffix, "-_")
```
I *think* (but double check me) that _ anywhere is considered invalid (not just beginning or end), so that may need removing or substituting, not just trimming from beginning or end

also, when we're trimming, it's then theoretically possible to have an empty string, so that case needs handling, I guess

aojea replied:
see https://github.com/kubernetes/kubernetes/pull/129790#discussion_r1931002836

aojea commented on staging/src/k8s.io/client-go/tools/record/util/util.go:
```diff
@@ -38,3 +43,26 @@ func IsKeyNotFoundError(err error) bool {
+
+var dns1123LabelNonValid = regexp.MustCompile("[^a-z0-9-.]")
+
+// GenerateEventName generates a valid Event name from the referenced name and the passed UNIX timestamp.
+// The referenced Object name may not be a valid name for Events and cause the Event to fail
+// to be created, so we need to sanitize the referenced name to be a valid Event name.
+// Ref: https://issues.k8s.io/127594
+func GenerateEventName(refName string, unixNano int64) string {
+	name := fmt.Sprintf("%s.%x", refName, unixNano)
+	if errs := apimachineryvalidation.NameIsDNSSubdomain(name, false); len(errs) > 0 {
+		suffix := fmt.Sprintf(".%x", unixNano) // it always ends in a digit
+		// a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.'
+		prefix := strings.ToLower(refName)
+		prefix = dns1123LabelNonValid.ReplaceAllString(prefix, "-")
+		// and must start and end with an alphanumeric character
+		prefix = strings.Trim(prefix, "-.")
+		if len(prefix)+len(suffix) > utilvalidation.DNS1123SubdomainMaxLength {
+			prefix = prefix[:utilvalidation.DNS1123SubdomainMaxLength-len(suffix)]
+		}
+		name = prefix + suffix
+	}
+	return name
```
also moved to this folder that is already imported by both implementations to avoid a possible risk of drifting

BenTheElder commented:
/triage accepted

dgrisonnet commented on staging/src/k8s.io/client-go/tools/record/event.go:
```diff
@@ -489,7 +489,7 @@ func (recorder *recorderImpl) makeEvent(ref *v1.ObjectReference, annotations map
 	return &v1.Event{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:        fmt.Sprintf("%v.%x", ref.Name, t.UnixNano()),
+			Name:        util.GenerateEventName(ref.Name, t.UnixNano()),
```
same with https://github.com/kubernetes/kubernetes/blob/d99d3f7eb734dc9284a90374fcc6d8cc32d1946d/staging/src/k8s.io/client-go/tools/record/events_cache.go#L52-L68

aojea replied:
I commented in https://github.com/kubernetes/kubernetes/pull/129790#discussion_r1936326934, creating the event name depending on the name itself leaves the code in a weird state that may also cause issues in the future as it can make it recursive if we are not careful, this approach is simpler and solves the problem

aojea replied:
those concatenation of strings does not guarantee the generate name is a valid name, see new approach

dgrisonnet commented on staging/src/k8s.io/client-go/tools/events/event_recorder.go:
```diff
@@ -96,7 +96,7 @@ func (recorder *recorderImpl) makeEvent(refRegarding *v1.ObjectReference, refRel
 	return &eventsv1.Event{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      fmt.Sprintf("%v.%x", refRegarding.Name, t.UnixNano()),
+			Name:      util.GenerateEventName(refRegarding.Name, t.UnixNano()),
```
I think we wanted to pass the event key here: https://github.com/kubernetes/kubernetes/blob/d99d3f7eb734dc9284a90374fcc6d8cc32d1946d/staging/src/k8s.io/client-go/tools/events/event_broadcaster.go#L58-L66 instead of refRegarding.Name + t.UnixNano() as we can be certain that the key is unique to each event

aojea replied:
I tried that and I discarded because I don't like the idea of generating the event object first and after generate the name,  
```
event := Event{} \\ without name
name := genKey(event)
event.Name = name
```

with this approach we have the same behavior as today plus we support the objects names that are not supported by the event name validation, it is more easier to think about, if we want unique key then just we can use a sha or uuid directly, no need to get the key

aojea replied:
@dgrisonnet the event key is not guaranteed a valid event.name

aojea commented:
/test

aojea commented on staging/src/k8s.io/client-go/tools/record/util/util.go:
```diff
@@ -38,3 +43,26 @@ func IsKeyNotFoundError(err error) bool {
+		prefix = strings.Trim(prefix, "-.")
+		if len(prefix)+len(suffix) > utilvalidation.DNS1123SubdomainMaxLength {
+			prefix = prefix[:utilvalidation.DNS1123SubdomainMaxLength-len(suffix)]
+		}
```
hmm, then I may take back my opinion and use the recommended option by @dgrisonnet , after playing abit with the logic the name generator starts to get more complex

liggitt reviewed (commented):
I like the simplification of just generating a random object name in invalid cases

aojea commented on staging/src/k8s.io/client-go/tools/record/util/util.go:
```diff
@@ -38,3 +42,37 @@ func IsKeyNotFoundError(err error) bool {
 	return statusErr != nil && statusErr.Status().Code == http.StatusNotFound
 }
+
+var dns1123LabelNonValid = regexp.MustCompile("[^a-z0-9-.]")
```
dropped

aojea commented on staging/src/k8s.io/client-go/tools/record/util/util.go:
```diff
@@ -38,3 +42,37 @@ func IsKeyNotFoundError(err error) bool {
+func GenerateEventName(refName string, unixNano int64) string {
+	name := fmt.Sprintf("%s.%x", refName, unixNano)
+	if errs := apimachineryvalidation.NameIsDNSSubdomain(name, false); len(errs) > 0 {
+		// Name must be a lowercase RFC 1123 subdomain, that  must consist of lower case alphanumeric
```
I didn't know those things existed, switched the code to use them

liggitt commented on staging/src/k8s.io/client-go/tools/record/util/util.go:
```diff
@@ -38,3 +41,27 @@ func IsKeyNotFoundError(err error) bool {
+		for _, char := range refName {
+			seed += int64(char) // char is a rune, which is an int32
+		}
+		apimachineryrand.Seed(seed)
+		name = fmt.Sprintf("%s.%x", apimachineryrand.String(length), unixNano)
```
not sure re-seeding and using a rand string generator that requires a global lock is a great idea... if the format produced by uuid.New() is valid for this, maybe just use that?

aojea replied:
it was just for making the unit test deterministic, the number of times this is going to be called is really small

liggitt replied:
hrmm... I don't love doing things for unit test determinism in a non test file

aojea replied:
uuid is fine, will switch to it then

aojea committed:
events: ensure the name is valid

The event Object is created from the referenced Object name, however,
there is no guarantee that the name from the referenced Object will be a
valid Event Name.

For those Objects with name not valid for an Event, generate a new valid
name using an UUID.

aojea commented on staging/src/k8s.io/client-go/tools/record/util/util.go:
```diff
@@ -38,3 +42,16 @@ func IsKeyNotFoundError(err error) bool {
+	name := fmt.Sprintf("%s.%x", refName, unixNano)
+	if errs := apimachineryvalidation.NameIsDNSSubdomain(name, false); len(errs) > 0 {
+		// Using an uuid guarantees uniqueness and correctness
+		name = uuid.New().String()
```
ping @liggitt

liggitt commented:
/lgtm
/approve
